src/leiningen/ring/uberwar.clj

```diff
@@ -13,25 +17,54 @@
     (get-cp project)
     (->> (:library-path project) io/file .listFiles (map str))))
 
-(defn contains-entry? [^java.io.File file ^String entry]
-  (with-open [jar-file (JarFile. file)]
-    (some (partial = entry)
-          (map #(.getName ^JarEntry %)
-               (enumeration-seq (.entries jar-file))))))
-
 (defn jar-dependencies [project]
   (for [pathname (get-classpath project)
-        :let [file (io/file pathname)
-              fname (.getName file)]
-        :when (and (.endsWith fname ".jar")
-                   ;; Servlet container will have it's own servlet-api impl
-                   (not (contains-entry? file "javax/servlet/Servlet.class")))]
-    file))
-
-(defn jar-entries [war project]
+        :when (str/ends-with? pathname ".jar")]
+    (io/file pathname)))
+
+(def ^:private javax-servlet? #{"javax/servlet/Servlet.class"})
+
+(defn- pom-properties? [file-name]
+  (str/ends-with? file-name "pom.properties"))
+
+(defn- read-jar-pom-properties [^String full-path]
+  (with-open [in (.openStream (URL. full-path))]
+    (->> (doto (Properties.) (.load in))
+         (into {}))))
+
+(defn- from-pom-properties
+  "Returns a jar name which includes the groupId, artifactId, & version,
```

Comment:
Docstrings of private functions should be omitted. Docstrings are for public APIs. If the purpose of a function isn't apparent from its name and content, it's an indication it should be refactored.